File: apps/mobile/lib/core/api_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:supabase_flutter/supabase_flutter.dart';

import '../config.dart';

class ApiException implements Exception {
  final int statusCode;
  final String message;

  ApiException(this.statusCode, this.message);

  bool get unauthorized => statusCode == 401;
  bool get forbidden => statusCode == 403;

  @override
  String toString() => message;
}

class ApiClient {
  final http.Client _client = http.Client();
  String? _token;

  String get _baseUrl => AppConfig.apiBaseUrl;

  String get baseUrl => _baseUrl;

  void updateToken(String? token) {
    _token = token;
  }

  String? get currentToken => _token;

  Map<String, String> _headers() => {
        'Content-Type': 'application/json',
        if (_token != null) 'Authorization': 'Bearer $_token',
      };

  Future<Map<String, dynamic>> get(String path) async {
    final uri = Uri.parse('$_baseUrl$path');
    print('[api] GET $uri');
    try {
      final res = await _client
          .get(uri, headers: _headers())
          .timeout(const Duration(seconds: 12));
      print('[api] GET $uri -> ${res.statusCode}');
      return _decode(res);
    } on TimeoutException {
      throw ApiException(
        0,
        'Could not reach the server. Check your connection or the API address.',
      );
    }
  }

  Future<Map<String, dynamic>> post(
    String path, {
    Map<String, dynamic>? body,
  }) async {
    final uri = Uri.parse('$_baseUrl$path');
    print('[api] POST $uri');
    try {
      final res = await _client
          .post(
            uri,
            headers: _headers(),
            body: body == null ? null : jsonEncode(body),
          )
          .timeout(const Duration(seconds: 12));
      print('[api] POST $uri -> ${res.statusCode}');
      return _decode(res);
    } on TimeoutException {
      throw ApiException(
        0,
        'Could not reach the server. Check your connection or the API address.',
      );
    }
  }

  Map<String, dynamic> _decode(http.Response res) {
    Map<String, dynamic>? json;
    try {
      json = jsonDecode(res.body) as Map<String, dynamic>;
    } catch (_) {
      json = null;
    }
    if (res.statusCode >= 200 && res.statusCode < 300) {
      return json ?? const {};
    }
    print('[api] ERROR ${res.statusCode}: ${res.body}');
    throw ApiException(
      res.statusCode,
      json?['error']?.toString() ?? 'Something went wrong (${res.statusCode}).',
    );
  }

  void dispose() => _client.close();
}

final ApiClient api = ApiClient();

void syncApiToken() {
  api.updateToken(Supabase.instance.client.auth.currentSession?.accessToken);
}

File: apps/mobile-rn/src/core/apiClient.ts
import { AppConfig } from '../config';

export class ApiException extends Error {
  statusCode: number;

  constructor(statusCode: number, message: string) {
    super(message);
    this.statusCode = statusCode;
  }

  get unauthorized(): boolean {
    return this.statusCode === 401;
  }

  get forbidden(): boolean {
    return this.statusCode === 403;
  }
}

const TIMEOUT_MS = 12000;

class ApiClient {
  private _token: string | null = null;

  get baseUrl(): string {
    return AppConfig.apiBaseUrl;
  }

  get currentToken(): string | null {
    return this._token;
  }

  updateToken(token: string | null): void {
    this._token = token;
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = { 'Content-Type': 'application/json' };
    if (this._token != null) h.Authorization = `Bearer ${this._token}`;
    return h;
  }

  private async request(
    method: 'GET' | 'POST',
    path: string,
    body?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const uri = `${this.baseUrl}${path}`;
    console.log(`[api] ${method} ${uri}`);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
    try {
      const res = await fetch(uri, {
        method,
        headers: this.headers(),
        body: body == null ? undefined : JSON.stringify(body),
        signal: controller.signal,
      });
      console.log(`[api] ${method} ${uri} -> ${res.status}`);
      return await this.decode(res);
    } catch (e: any) {
      if (e?.name === 'AbortError') {
        throw new ApiException(
          0,
          'Could not reach the server. Check your connection or the API address.',
        );
      }
      throw new ApiException(
        0,
        'Could not reach the server. Check your connection or the API address.',
      );
    } finally {
      clearTimeout(timer);
    }
  }

  private async decode(res: Response): Promise<Record<string, unknown>> {
    let json: Record<string, unknown> | null = null;
    try {
      json = (await res.json()) as Record<string, unknown>;
    } catch {
      json = null;
    }
    if (res.status >= 200 && res.status < 300) {
      return json ?? {};
    }
    console.log(`[api] ERROR ${res.status}: ${res.statusText}`);
    const message = json?.['error']
      ? String(json['error'])
      : `Something went wrong (${res.status}).`;
    throw new ApiException(res.status, message);
  }

  async get(path: string): Promise<Record<string, unknown>> {
    return this.request('GET', path);
  }

  async post(
    path: string,
    body?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    return this.request('POST', path, body);
  }
}

export const api = new ApiClient();